src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,216 @@
+            tail = newNode;
+            size++;
+            return;
+        }
+        ListNode temp = findElementByIndex(index);
+        newNode.setNext(temp);
+        newNode.setPrev(temp.getPrev());
+        temp.setPrev(newNode);
+        newNode.getPrev().setNext(newNode);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        ListNode temp = findElementByIndex(index);
+        return (T) temp.getValue();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        ListNode temp = findElementByIndex(index);
+        T oldValue = (T) temp.getValue();
+        temp.setValue(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        ListNode temp = findElementByIndex(index);
+        unlink(temp);
+        return (T) temp.getValue();
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (size == 0) {
+            return false;
+        }
+        boolean found = false;
+        ListNode temp = head;
+        if (temp.getValue() != null && temp.getValue().equals(object)
+                || temp.getValue() == object) {
+            unlink(head);
+            found = true;
+            return found;
+        }
+        while (temp != tail && !found) {
+            if (temp.getValue() != null && temp.getValue().equals(object)
+                    || temp.getValue() == object) {
+                unlink(temp);
+                found = true;
+            }
+            if (found) {
+                return found;
+            }
+            temp = temp.getNext();
+        }
+        if (temp.getValue() != null && temp.getValue().equals(object)
+                || temp.getValue() == object) {
+            unlink(tail);
+            found = true;
+        }
+        return found;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private ListNode findElementByIndex(int index) {
+        if (index < 0 || index > size - 1 || isEmpty()) {
+            throw new IndexOutOfBoundsException("invalid index " + index);
+        }
+        if (index == 0) {
+            return head;
+        }
+        if (index == size - 1) {
+            return tail;
+        }
+        if (index > size / 2) {
+            ListNode temp = tail;
+            index = size - index - 1;
+            while (index != 0) {
+                temp = temp.getPrev();
+                index--;
+            }
+            return temp;
+        }
+        ListNode temp = head;
+        while (index != 0) {
+            temp = temp.getNext();
+            index--;
+        }
+        return temp;
```

Comment:
use two for-loops, first from head to tail, second from tail to head, and switch between them

Reply:
this will double cost of operation here cause of doubling pointers operartions

Reply:
and complicating conditions. disagree. resolve please without change

Reply:
i serach here from end or start depending which is closer to index.